File: 0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.py
class Solution:
        # @param {TreeNode} root
        # @return {integer}
        def leftheight(self, node):
            height = 0
            while(node):
                height = height +1
                node = node.left
            return height
        
        def rightheight(self, node):
            height = 0
            while(node):
                height = height +1
                node = node.right
            return height
        def countNodes(self, root):
            if(root == 0):
                return 0
            lh = self.leftheight(root)
            rh = self.rightheight(root)
            if(lh == rh):
                return pow(2,lh)-1
            else:
                return 1+self.countNodes(root.left)+self.countNodes(root.right)
        

File: 0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        if(root == NULL){
            return 0;
        }
        
        int lh = leftHeight(root);
        int rh = rightHeight(root);
    
        if(lh==rh){
            return (1<<lh) - 1;
        }
        else{
        return 1 + countNodes(root->left) + countNodes(root->right);
        }
    }
    int leftHeight(TreeNode* node){
        int height = 0;
        while(node){
            height ++;
            node = node->left;
        }
        return height;
    }
        int rightHeight(TreeNode* node){
        int height = 0;
        while(node){
            height ++;
            node = node->right;
        }
        return height;
    }
    
};

//TC : O(logN)^2